api/submission/fetch_submission_history.go

```diff
@@ -0,0 +1,49 @@
+package submission
+
+import (
+    "github.com/eriq-augustine/autograder/api/core"
+    "github.com/eriq-augustine/autograder/db"
+    "github.com/eriq-augustine/autograder/model"
+)
+
+type FetchSubmissionHistoryRequest struct {
+    core.APIRequestAssignmentContext
+    core.MinRoleGrader
+
+    TargetUser       core.TargetUserSelfOrGrader `json:"target-email"`
+}
+
+type FetchSubmissionHistoryResponse struct {
+    FoundUser        bool `json:"found-user"`
+    FoundSubmissions bool `json:"found-submissions"`
+    GradingResults []*model.GradingResult `json:"grading-results"`
+}
+
+func HandleFetchSubmissionHistory(request *FetchSubmissionHistoryRequest) (*FetchSubmissionHistoryResponse, *core.APIError) {
+    response := FetchSubmissionHistoryResponse{
+        FoundUser:        false,
+        FoundSubmissions: false,
```

Comment:
The semantics around FoundSubmissions are a bit strange and rely on knowledge of the DB state.

If the user has submission(s), the this will be true and the results will be non-empty.
(Makes sense.)

If the user has never made a submission, then this will be true and the results will be nil.
(Pretty strange.)

If the user made a submission that was removed (the top-level submission dir exists with no contents), then this will be true and the results will be an empty list.
(Also pretty strange.)

I think the thing to do is remove FoundSubmissions and return an empty slice from the DB when there are no submissions, but no errors.
(You see this behavior a lot with containers: return nil on error, and empty on missing cases.)

Reply:
I think I am misunderstanding the cases here.  Can you help me understand why it would be true in case 2 and 3?  From my understanding, if a user has never made a submission or if they have no directory entries it will return nil, nil from:

```go
func (this *backend) GetSubmissionResultHistory(assignment *model.Assignment, email string) ([]*model.GradingResult, error) {
    submissions := make([]*model.GradingResult, 0);

    submissionsDir := this.getUserSubmissionDir(assignment.GetCourse().GetID(), assignment.GetID(), email);
    if !util.PathExists(submissionsDir) {
        return nil, nil;
    }

    dirents, err := os.ReadDir(submissionsDir);
    if err != nil {
        return nil, fmt.Errorf("Unable to read user submissions dir '%s': '%w'.", submissionsDir, err);
    }

    if len(dirents) == 0 {
        return nil, nil;
    }

    for _, dirent := range dirents {
        submission, err := this.GetSubmissionContents(assignment, email, dirent.Name());
        if err != nil {
            return nil, fmt.Errorf("Unable to get submission contents for %s: .", email);
        }

        submissions = append(submissions, submission);
    }

    return submissions, nil;
}
``` 
Thus entering the if gradingResult == nil block here:

```go
gradingResults, err := db.GetSubmissionResultHistory(request.Assignment, request.TargetUser.Email);
    if err != nil {
        return nil, core.NewInternalError("-607", &request.APIRequestCourseUserContext, "Failed to get submissions.").
            Err(err).Add("email", request.TargetUser.Email);
    }

    if gradingResults == nil {
        return &response, nil;
    }

    response.FoundSubmissions = true;
    response.GradingResults = gradingResults;

    return &response, nil;
}
``` 
Which would return an empty slice and no error.  I have no issue removing FoundSubmissions from the implementation, I just want to make sure I understand!  Thank you!

Reply:
In my most recent commit I have removed FoundSubmissions and am returning an empty slice for cases 2 and 3.